pilot/pkg/model/push_context.go

```diff
@@ -985,6 +1048,61 @@ func (ps *PushContext) initDestinationRules(env *Environment) error {
 	return nil
 }
 
+// AuthenticationPolicyForWorkload returns the matching auth policy for a given service
+// This replaces store.AuthenticationPolicyForWorkload
+func (ps *PushContext) AuthenticationPolicyForWorkload(service *Service, l labels.Instance, port *Port) *authn.Policy {
+	var workloadPolicy *authn.Policy
+
+	// Match by Service hostname
+	workloadPolicy = authenticationPolicyForWorkload(ps.AuthNPolicies[service.Hostname], service, l, port)
+
+	// Match by namespace
+	if workloadPolicy == nil {
+		workloadPolicy = authenticationPolicyForWorkload(ps.AuthNPolicies[host.Name(service.Attributes.Namespace)], service, l, port)
+	}
+
+	// Use default global authentication policy
+	if workloadPolicy == nil {
+		workloadPolicy = ps.defaultAuthNMeshPolicy
+	}
+
+	return workloadPolicy
+}
+
+func authenticationPolicyForWorkload(specsByPort map[int][]*Config, service *Service, l labels.Instance, port *Port) *authn.Policy {
+	var matchedPolicy *authn.Policy
+	if specsByPort == nil {
+		return matchedPolicy
+	}
+	if matchedPolicy = matchPolicyByLabels(specsByPort[port.Port], service, l); matchedPolicy == nil {
```

Comment:
@diemtvu what is Istio's policy on this?  Because the existing validation logic has nothing to stop someone from creating creating labels (in fact there is no validation for them whatsoever):
https://github.com/istio/istio/blob/master/pkg/config/validation/validation.go#L1733

Labels are currently getting passed in to the existing XDS logic here:
https://github.com/istio/istio/blob/master/pilot/pkg/security/authn/v1alpha1/model.go#L31
And the existing `istioConfigStore.AuthenticationPolicyForWorkload` lookup is properly filtering polices on them here:
https://github.com/istio/istio/blob/master/pilot/pkg/model/config.go#L571

TargetSelector.Labels are currently [hidden](https://github.com/istio/api/blob/master/authentication/v1alpha1/policy.pb.go#L1014) in docs and on istio.io.

To summarize, My existing implementation is such that feature parity is maintained with `master`.  Filtering auth policies by labels appears to be implemented in code for the past N releases - can I do away with that function/feature since it's not officially documented or does it require an additional approval process?

cc @linsun @rshriram

Reply:
I guess the question is, what is the scope of this PR?
- performance improvement?  i.e. functionally equivalent logic but with a cache to minimize lookup calls
OR
- above item + restructure of the lookup code for the incoming label matching feature down the line

Reply:
My point is this implementation is not correct with the intention for the label. On the other hand, the exiting label selector could be premature, and while we have it in API (my bad), we haven't officially announce supporting for it. That's why I think it's possible to ignore it for now. If we want to support it here, then we have to do it the right way. (see next)

TL;DR. The label selector is supposed to replace service (host name) selector. That is to say, if a policy use a label selector, it will be used to matched with workload and *ignore* host-name match. Also, the labels are matched with the server workload (aka destination service instance).

The way we implement in this PR so far is constructing a map from host + port -> policy. You then try to match the host name first, then filter further on label. This is not the same. The part that construct the map (`initAuthNPolicies`) also assumes the hostname is set. To have the same behavior as we have on the server side, you need to iterate all the policies (on the server's namespace) and try to match with label first; if nothing found, then match on the destination hostname.

We also want to make sure the listener builder code follow the same mechanism to select policy, otherwise, there could be a bug due to discrepancy between client and server side. That's why I insist to share the code between two as much as possible.

Reply:
Reading through the thread, I see mainly two points

- Label selector authn policy. I agree that current label selector part in this PR is incorrect w.r.t the API's intention. However
  > above item + restructure of the lookup code for the incoming label matching feature down the line

what I understand @diemtvu is that we don't need to worry about the label-selected policy at all, we deal with that later. This PR we just remove the handling of the label selector part. Pretty much we just need to remove `matchPolicyByLabels`, it'll just make the PR smaller.

- Code duplication part. Note that caching the policy in the push context is similar to other CRD, (DestinationRule), and indeed this remove the need of re-calculating the policy in the CDS code every time, the for loop. This makes a lot of sense. The concern is then we need to have one impl/lib for listeners code. I think the push context can be used for listeners.go as well?

The alternative to use original config.AuthenticationPolicyForWorkload(), then we need to have the same optimization implmemented as push context for pre-caching. The mode/config part is not tied with the push init/new push lifecycle, and require quite some additional work. So I think adding to push context and switch to the same caching in the `listener.go` looks like a legit plan.